npm/src/language-service/language-service.ts

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import type {
+  IDiagnostic,
+  ICompletionList,
+  IHover,
+  IDefinition,
+  LanguageService,
+} from "../../lib/node/qsc_wasm.cjs";
+import { log } from "../log.js";
+import {
+  VSDiagnostic,
+  mapDiagnostics,
+  mapUtf16UnitsToUtf8Units,
+  mapUtf8UnitsToUtf16Units,
+} from "../vsdiagnostic.js";
+import { IServiceEventTarget, IServiceProxy } from "../worker-proxy.js";
+type QscWasm = typeof import("../../lib/node/qsc_wasm.cjs");
+
+// Only one event type for now
+export type LanguageServiceEvent = {
+  type: "diagnostics";
+  detail: {
+    uri: string;
+    version: number;
+    diagnostics: VSDiagnostic[];
+  };
+};
+
+// These need to be async/promise results for when communicating across a WebWorker, however
+// for running the compiler in the same thread the result will be synchronous (a resolved promise).
+export interface ILanguageService {
+  updateDocument(uri: string, version: number, code: string): Promise<void>;
+  closeDocument(uri: string): Promise<void>;
+  getCompletions(documentUri: string, offset: number): Promise<ICompletionList>;
+  getHover(documentUri: string, offset: number): Promise<IHover | null>;
+  getDefinition(
+    documentUri: string,
+    offset: number
+  ): Promise<IDefinition | null>;
+  dispose(): Promise<void>;
+
+  addEventListener<T extends LanguageServiceEvent["type"]>(
```

Comment:
Are addEventListener and removeEventListener proxied to the worker also? Do they not need to return a Promise if so like the other APIs on the proxies?

Reply:
No, these two get special treatment because they're event methods. The proxy will listen to event messages coming from the worker thread and forward them to the subscribers on the main thread.

Here, I attempted to list the different types of methods a proxyable "service" can have.

https://github.com/microsoft/qsharp/blob/bec69c3410cec5461b8f52da14810eab00e863f5/npm/src/worker-proxy.ts#L20

Snippet from above link:

```ts

/**
 * Defines the service methods that the proxy will handle and their types.
 * "request" is a normal async method.
 * "requestWithProgress" methods take an @see IServiceEventTarget to
 *   communicate events back to the main thread as they run. They also set
 *   the service state to "busy" while they run.
 * "addEventListener" and "removeEventListener" methods are used to
 *   subscribe to events from the service.
 */
export type MethodMap<T> = {
  [M in keyof T]:
    | "request"
    | "requestWithProgress"
    | "addEventListener"
    | "removeEventListener";
};
```